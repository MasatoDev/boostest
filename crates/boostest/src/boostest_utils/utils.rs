use anyhow::{anyhow, Result};
use std::path::{Path, PathBuf};

use glob::glob;
use std::collections::HashMap;
use std::ffi::OsStr;
use std::fs::{self, File};
use std::io::{self, prelude::*};

pub fn read(path: &Path) -> io::Result<String> {
    let mut f = File::open(path)?;
    let mut s = String::new();
    match f.read_to_string(&mut s) {
        Ok(_) => Ok(s),
        Err(e) => Err(e),
    }
}

pub fn read_matching_files(
    patterns: Vec<String>,
    out_file_name: &String,
) -> anyhow::Result<HashMap<PathBuf, String>> {
    let mut contents: HashMap<PathBuf, String> = HashMap::new();

    for pattern in &patterns {
        let result_glob = glob(pattern)?;

        for entry in result_glob {
            let path = entry?;

            if path
                .file_name()
                .unwrap_or(OsStr::new("none_file_name"))
                .to_string_lossy()
                .contains(out_file_name.as_str())
            {
                continue;
            }

            if path.is_file() {
                contents.insert(path.clone(), fs::read_to_string(path)?);
            }
        }
    }

    Ok(contents)
}

pub fn find_boostest_json_recursive(dir: PathBuf) -> anyhow::Result<PathBuf> {
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() {
            if let Some(file_name) = path.file_name() {
                if file_name == "boostest.setting.json" {
                    return Ok(path);
                }
            }
        } else if path.is_dir() {
            if let Ok(found_path) = find_boostest_json_recursive(path) {
                return Ok(found_path);
            }
        }
    }

    Err(anyhow!("boostest.json not found"))
}

/**
 * Normalize and resolve the path.
 * parent: "hoge/a/b/c", relative: "../../d/e/f" => "hoge/a/d/e/f"
 */
pub fn normalize_and_resolve_path(parent: &Path, relative: &Path) -> Result<PathBuf, String> {
    let combined_path = parent.join(relative);
    let resolved_path = combined_path
        .components()
        .fold(PathBuf::new(), |mut acc, comp| {
            match comp {
                std::path::Component::ParentDir => {
                    acc.pop();
                }
                std::path::Component::CurDir => {}
                other => acc.push(other),
            }
            acc
        });

    if resolved_path.is_absolute() {
        Ok(resolved_path)
    } else {
        Err("The resolved path is not absolute.".to_string())
    }
}
