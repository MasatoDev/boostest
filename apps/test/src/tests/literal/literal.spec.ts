import { runSnapshotTest, failedTest } from "../utils";
import {
  // string
  TsTypeLiteralString,
  TsTypeLiteralLiteralTypeString,
  TsLiteralTypeStringUnionType,

  // number
  TsTypeLiteralNumber,
  TsTypeLiteralLiteralNumberType,
  TsLiteralNumberUnionType,

  // boolean
  TsTypeLiteralBoolean,
  TsTypeLiteralLiteralBooleanType,
  TsLiteralBooleanUnionType,

  // null
  TsTypeLiteralNull,

  // undefined
  TsTypeLiteralUndefined,

  // Array
  TsTypeLiteralArray,
  TsTypeLiteralLiteralArrayType,
  TsLiteralArrayUnionType,

  // Object
  TsTypeLiteralObject,
  TsTypeLiteralLiteralObjectType,
  TsLiteralObjectUnionType,

  // Function
  TsTypeLiteralFunction,
  TsTypeLiteralLiteralFunctionType,
  TsLiteralFunctionUnionType,

  // Symbol
  TsTypeLiteralSymbol,
  TsTypeLiteralLiteralSymbolType,

  // use
  UseTSTypeLiteralAlias,
  UseTSTypeLiteralInterface,
} from "@/ts_types/ts_type_literal";
import { boostestTsTypeLiteralString } from "./boostest_output/boostestTsTypeLiteralString";
import { boostestTsTypeLiteralLiteralTypeString } from "./boostest_output/boostestTsTypeLiteralLiteralTypeString";
import { boostestTsLiteralTypeStringUnionType } from "./boostest_output/boostestTsLiteralTypeStringUnionType";
import { boostestTsTypeLiteralNumber } from "./boostest_output/boostestTsTypeLiteralNumber";
import { boostestTsTypeLiteralLiteralNumberType } from "./boostest_output/boostestTsTypeLiteralLiteralNumberType";
import { boostestTsLiteralNumberUnionType } from "./boostest_output/boostestTsLiteralNumberUnionType";
import { boostestTsTypeLiteralBoolean } from "./boostest_output/boostestTsTypeLiteralBoolean";
import { boostestTsTypeLiteralLiteralBooleanType } from "./boostest_output/boostestTsTypeLiteralLiteralBooleanType";
import { boostestTsLiteralBooleanUnionType } from "./boostest_output/boostestTsLiteralBooleanUnionType";
import { boostestTsTypeLiteralNull } from "./boostest_output/boostestTsTypeLiteralNull";
import { boostestTsTypeLiteralUndefined } from "./boostest_output/boostestTsTypeLiteralUndefined";
import { boostestUseTSTypeLiteralInterface } from "./boostest_output/boostestUseTSTypeLiteralInterface";
import { boostestUseTSTypeLiteralAlias } from "./boostest_output/boostestUseTSTypeLiteralAlias";
import { boostestTsTypeLiteralSymbol } from "./boostest_output/boostestTsTypeLiteralSymbol";
import { boostestTsLiteralFunctionUnionType } from "./boostest_output/boostestTsLiteralFunctionUnionType";
import { boostestTsTypeLiteralLiteralFunctionType } from "./boostest_output/boostestTsTypeLiteralLiteralFunctionType";
import { boostestTsTypeLiteralFunction } from "./boostest_output/boostestTsTypeLiteralFunction";
import { boostestTsLiteralObjectUnionType } from "./boostest_output/boostestTsLiteralObjectUnionType";
import { boostestTsTypeLiteralLiteralObjectType } from "./boostest_output/boostestTsTypeLiteralLiteralObjectType";
import { boostestTsTypeLiteralObject } from "./boostest_output/boostestTsTypeLiteralObject";
import { boostestTsLiteralArrayUnionType } from "./boostest_output/boostestTsLiteralArrayUnionType";
import { boostestTsTypeLiteralLiteralArrayType } from "./boostest_output/boostestTsTypeLiteralLiteralArrayType";
import { boostestTsTypeLiteralArray } from "./boostest_output/boostestTsTypeLiteralArray";

describe("Literal Type Tests", () => {
  runSnapshotTest(
    "TsTypeLiteralString",
    boostestTsTypeLiteralString<TsTypeLiteralString>(),
  );
  runSnapshotTest(
    "TsTypeLiteralLiteralTypeString",
    boostestTsTypeLiteralLiteralTypeString<TsTypeLiteralLiteralTypeString>(),
  );
  runSnapshotTest(
    "TsLiteralTypeStringUnionType",
    boostestTsLiteralTypeStringUnionType<TsLiteralTypeStringUnionType>(),
  );
  runSnapshotTest(
    "TsTypeLiteralNumber",
    boostestTsTypeLiteralNumber<TsTypeLiteralNumber>(),
  );
  runSnapshotTest(
    "TsTypeLiteralLiteralNumberType",
    boostestTsTypeLiteralLiteralNumberType<TsTypeLiteralLiteralNumberType>(),
  );
  runSnapshotTest(
    "TsLiteralNumberUnionType",
    boostestTsLiteralNumberUnionType<TsLiteralNumberUnionType>(),
  );
  runSnapshotTest(
    "TsTypeLiteralBoolean",
    boostestTsTypeLiteralBoolean<TsTypeLiteralBoolean>(),
  );
  runSnapshotTest(
    "TsTypeLiteralLiteralBooleanType",
    boostestTsTypeLiteralLiteralBooleanType<TsTypeLiteralLiteralBooleanType>(),
  );
  runSnapshotTest(
    "TsLiteralBooleanUnionType",
    boostestTsLiteralBooleanUnionType<TsLiteralBooleanUnionType>(),
  );
  runSnapshotTest(
    "TsTypeLiteralNull",
    boostestTsTypeLiteralNull<TsTypeLiteralNull>(),
  );
  runSnapshotTest(
    "TsTypeLiteralUndefined",
    boostestTsTypeLiteralUndefined<TsTypeLiteralUndefined>(),
  );
  runSnapshotTest(
    "TsTypeLiteralArray",
    boostestTsTypeLiteralArray<TsTypeLiteralArray>(),
  );
  runSnapshotTest(
    "TsTypeLiteralLiteralArrayType",
    boostestTsTypeLiteralLiteralArrayType<TsTypeLiteralLiteralArrayType>(),
  );
  runSnapshotTest(
    "TsLiteralArrayUnionType",
    boostestTsLiteralArrayUnionType<TsLiteralArrayUnionType>(),
  );
  runSnapshotTest(
    "TsTypeLiteralObject",
    boostestTsTypeLiteralObject<TsTypeLiteralObject>(),
  );
  runSnapshotTest(
    "TsTypeLiteralLiteralObjectType",
    boostestTsTypeLiteralLiteralObjectType<TsTypeLiteralLiteralObjectType>(),
  );
  runSnapshotTest(
    "TsLiteralObjectUnionType",
    boostestTsLiteralObjectUnionType<TsLiteralObjectUnionType>(),
  );
  runSnapshotTest(
    "TsTypeLiteralFunction",
    boostestTsTypeLiteralFunction<TsTypeLiteralFunction>(),
  );
  runSnapshotTest(
    "TsTypeLiteralLiteralFunctionType",
    boostestTsTypeLiteralLiteralFunctionType<TsTypeLiteralLiteralFunctionType>(),
  );
  runSnapshotTest(
    "TsLiteralFunctionUnionType",
    boostestTsLiteralFunctionUnionType<TsLiteralFunctionUnionType>(),
  );
  runSnapshotTest(
    "TsTypeLiteralSymbol",
    boostestTsTypeLiteralSymbol<TsTypeLiteralSymbol>(),
  );
  runSnapshotTest(
    "TsTypeLiteralAlias",
    boostestUseTSTypeLiteralAlias<UseTSTypeLiteralAlias>(),
  );
  runSnapshotTest(
    "TsTypeLiteralInterface",
    boostestUseTSTypeLiteralInterface<UseTSTypeLiteralInterface>(),
  );

  // failedTest("TsTypeLiteralLiteralSymbol");
});
