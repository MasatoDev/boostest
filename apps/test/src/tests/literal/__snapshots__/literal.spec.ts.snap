// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Literal Type Tests TsLiteralArrayUnionType matches snapshot 1`] = `"[]"`;

exports[`Literal Type Tests TsLiteralBooleanUnionType matches snapshot 1`] = `"false"`;

exports[`Literal Type Tests TsLiteralFunctionUnionType matches snapshot 1`] = `
"(x) => {
        return 10;
    }"
`;

exports[`Literal Type Tests TsLiteralNumberUnionType matches snapshot 1`] = `"1"`;

exports[`Literal Type Tests TsLiteralObjectUnionType matches snapshot 1`] = `"{"type":"A"}"`;

exports[`Literal Type Tests TsLiteralTypeStringUnionType matches snapshot 1`] = `""A""`;

exports[`Literal Type Tests TsTypeLiteralAlias matches snapshot 1`] = `"{"literalString":"test string data","literalLiteralString":"string","literalStringUnion":"A","literalNumber":10,"literalLiteralNumber":42,"literalNumberUnion":1,"literalBoolean":false,"literalLiteralBoolean":true,"literalBooleanUnion":false,"literalNull":null,"literalUndefined":"undefined","literalArray":[],"literalLiteralArray":[1,2,3],"literalArrayUnion":[],"literalObject":{},"literalLiteralObject":{"name":"test string data","age":10},"literalObjectUnion":{"type":"A"},"literalFunction":"() => { }","literalLiteralFunction":"() => { }","literalFunctionUnion":"(x) => {\\n            return 10;\\n        }","literalBigInt":"9007199254740991n","literalLiteralBigInt":"123n","literalBigIntUnion":"1n","literalSymbol":"Symbol()"}"`;

exports[`Literal Type Tests TsTypeLiteralArray matches snapshot 1`] = `"[]"`;

exports[`Literal Type Tests TsTypeLiteralBoolean matches snapshot 1`] = `"false"`;

exports[`Literal Type Tests TsTypeLiteralFunction matches snapshot 1`] = `"() => { }"`;

exports[`Literal Type Tests TsTypeLiteralInterface matches snapshot 1`] = `"{"literalString":"test string data","literalLiteralString":"string","literalStringUnion":"A","literalNumber":10,"literalLiteralNumber":42,"literalNumberUnion":1,"literalBoolean":false,"literalLiteralBoolean":true,"literalBooleanUnion":false,"literalNull":null,"literalUndefined":"undefined","literalArray":[],"literalLiteralArray":[1,2,3],"literalArrayUnion":[],"literalObject":{},"literalLiteralObject":{"name":"test string data","age":10},"literalObjectUnion":{"type":"A"},"literalFunction":"() => { }","literalLiteralFunction":"() => { }","literalFunctionUnion":"(x) => {\\n            return 10;\\n        }","literalBigInt":"9007199254740991n","literalLiteralBigInt":"123n","literalBigIntUnion":"1n","literalSymbol":"Symbol()"}"`;

exports[`Literal Type Tests TsTypeLiteralLiteralArrayType matches snapshot 1`] = `"[1,2,3]"`;

exports[`Literal Type Tests TsTypeLiteralLiteralBooleanType matches snapshot 1`] = `"true"`;

exports[`Literal Type Tests TsTypeLiteralLiteralFunctionType matches snapshot 1`] = `"() => { }"`;

exports[`Literal Type Tests TsTypeLiteralLiteralNumberType matches snapshot 1`] = `"42"`;

exports[`Literal Type Tests TsTypeLiteralLiteralObjectType matches snapshot 1`] = `"{"name":"test string data","age":10}"`;

exports[`Literal Type Tests TsTypeLiteralLiteralTypeString matches snapshot 1`] = `""string""`;

exports[`Literal Type Tests TsTypeLiteralNull matches snapshot 1`] = `"null"`;

exports[`Literal Type Tests TsTypeLiteralNumber matches snapshot 1`] = `"10"`;

exports[`Literal Type Tests TsTypeLiteralObject matches snapshot 1`] = `"{}"`;

exports[`Literal Type Tests TsTypeLiteralString matches snapshot 1`] = `""test string data""`;

exports[`Literal Type Tests TsTypeLiteralSymbol matches snapshot 1`] = `"Symbol()"`;

exports[`Literal Type Tests TsTypeLiteralUndefined matches snapshot 1`] = `""undefined""`;
